model/riscv_insts_crypto.sail

```diff
@@ -0,0 +1,185 @@
+mapping clause assembly = SHA256SIG0 (rs1,rd) <-> "sha256sig0" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+mapping clause assembly = SHA256SIG1 (rs1,rd) <-> "sha256sig1" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+mapping clause assembly = SHA256SUM0 (rs1,rd) <-> "sha256sum0" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+mapping clause assembly = SHA256SUM1 (rs1,rd) <-> "sha256sum1" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+
+val      crypto_sha256 : (sha256_op, regidx, regidx) -> Retired effect {escape,rreg,wreg}
+/* --- specification snippet begin --- */
+function crypto_sha256   (op        , rd    , rs1   ) =  {
+  let inb    : bits(32) = X(rs1)[31..0];
+  let result : bits(32) = match op {
+    OP_SHA256_SIG0 => ror32(inb ,07) ^ ror32(inb ,18) ^ (inb >>  3),
+    OP_SHA256_SIG1 => ror32(inb ,17) ^ ror32(inb ,19) ^ (inb >> 10),
+    OP_SHA256_SUM0 => ror32(inb , 2) ^ ror32(inb, 13) ^ ror32(inb,22),
+    OP_SHA256_SUM1 => ror32(inb , 6) ^ ror32(inb, 11) ^ ror32(inb,25)
+  };
+  X(rd) = EXTZ(result);
+  RETIRE_SUCCESS
+}
+/* --- specification snippet end --- */
+
+/* Execution clauses for the SHA256 instructions. */
+function clause execute (SHA256SIG0 (rs1,rd)) = {
+    crypto_sha256(OP_SHA256_SIG0, rd, rs1);
+}
+
+function clause execute (SHA256SIG1 (rs1,rd)) = {
+    crypto_sha256(OP_SHA256_SIG1, rd, rs1);
+}
+
+function clause execute (SHA256SUM0 (rs1,rd)) = {
+    crypto_sha256(OP_SHA256_SUM0, rd, rs1);
+}
+
+function clause execute (SHA256SUM1 (rs1,rd)) = {
+    crypto_sha256(OP_SHA256_SUM1, rd, rs1);
+}
+
+/*
+ * Scalar Cryptography Extension - Scalar        SM3 instructions
+ * ----------------------------------------------------------------------
+ */
+
+
+/* AST Types for SM3.p0/1 */
+union   clause ast      =  SM3P0 : (regidx,regidx)
+union   clause ast      =  SM3P1 : (regidx,regidx)
+
+/* Decode clause for sm3.p0 */
+mapping clause encdec =           SM3P0   (rs1,rd) 
+    <-> 0b00 @ 0b01000 @ 0b01000 @ rs1 @ 0b001 @ rd @ 0b0010011
+
+/* Decode clause for sm3.p1 */
+mapping clause encdec =           SM3P1   (rs1,rd) 
+    <-> 0b00 @ 0b01000 @ 0b01001 @ rs1 @ 0b001 @ rd @ 0b0010011
+
+
+/* Asssembly code mapping for sm3p0 */
+mapping clause assembly =  SM3P0 (rs1,rd) <->
+    "sm3.p0" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+
+/* Asssembly code mapping for sm3p1 */
+mapping clause assembly =  SM3P1 (rs1,rd) <->
+    "sm3.p1" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+
+
+val      crypto_sm3 : (sm3_op, regidx, regidx) -> Retired effect {escape,rreg,wreg}
+/* --- specification snippet begin --- */
+function crypto_sm3   (op     , rd    , rs1   ) =  {
+  let r1    : bits(32) = X(rs1)[31..0];
+  let result  : bits(32) = match op {
+    P0  => r1 ^ rol32(r1 , 9) ^ rol32(r1 ,17),
+    P1  => r1 ^ rol32(r1 ,15) ^ rol32(r1 ,23)
+  };
+  X(rd)       = EXTZ(result);
+  RETIRE_SUCCESS
+}
+/* --- specification snippet end --- */
+
+
+/* Execute clause for sm3.p0 */
+function clause execute ( SM3P0 (rs1,rd)) = {
```

Comment:
Spacing

Reply:
For the avoidance of doubt:
```
  X(rd) = EXTZ(result);
```
is what you're after?

Reply:
Not sure what line you're talking about; this comment shows on:
```
function clause execute ( SM3P0 (rs1,rd)) = {
```
for me which should be:
```
function clause execute (SM3P0(rs1, rd)) = {
```